### PR TITLE
Tidy up ViewState collection from ViewModels

### PR DESCRIPTION
**Pull request by chrisbanes**
Using `flowWithLifecycle()` to reduce resource usage when the app is put into the background

### CODE REVIEW

File: common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt [hunk 1]
```diff
@@ -33,6 +36,23 @@ class SnackbarManager @Inject constructor() {
     private val pendingErrors = MutableSharedFlow<UiError>(extraBufferCapacity = 3)
     private val removeErrorSignal = MutableSharedFlow<Unit>(extraBufferCapacity = 1)
 
+    val flow: Flow<UiError?> = flow {
```
Comment (manuelvicnt) [inline common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt:39]: Can you add comments about what's the purpose of this?

File: common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt [hunk 2]
```diff
@@ -33,6 +36,23 @@ class SnackbarManager @Inject constructor() {
     private val pendingErrors = MutableSharedFlow<UiError>(extraBufferCapacity = 3)
```
Comment (manuelvicnt) [inline common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt:36]: Does this work as you expect? Even if there are no collectors, messages will be sent and they could be potentially lost. This `extraBufferCapacity` is only helpful when `SharedFlow` has slow subscribers. `emit()` will suspend until all collectors receive the event.

File: ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt [hunk 3]
```diff
@@ -55,29 +56,38 @@ internal class EpisodeDetailsViewModel @Inject constructor(
...
-        viewModelScope.launch {
-            observeEpisodeWatches.observe()
-                .onStart { emit(emptyList()) }
-                .collect { updateFromEpisodeWatches(it) }
-        }
+    val state = combine(
+        observeEpisodeDetails.observe(),
+        observeEpisodeWatches.observe(),
+        loadingState.observable,
+        snackbarManager.flow.onStart { emit(null) },
```
Comment (manuelvicnt) [inline ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt:71]: Why do you need to emit `null` on start? `snackbarManager.flow` already does that for you, right?

File: ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt [hunk 4]
```diff
@@ -55,29 +56,38 @@ internal class EpisodeDetailsViewModel @Inject constructor(
...
+            watches = episodeWatches,
+            canAddEpisodeWatch = episodeDetails.episode?.firstAired?.let {
+                it.isBefore(OffsetDateTime.now())
+            } ?: true,
+            refreshing = refreshing,
+            error = error,
+        )
+    }.stateIn(
+        scope = viewModelScope,
+        started = WhileSubscribed(5000),
```
Comment (manuelvicnt) [inline ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt:86]: This is only helpful if you're collecting the flow using the new `addRepeatingJob` API or you're manually cancelling the job/scope when the view goes to the background. This won't work if you're using `launchWhenStarted`.

Which one are you using?

Comment (chrisbanes) [reply ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt:86]: Hmmm, these are collected from Compose via `collectAsState()`. Oh well.

Comment (manuelvicnt) [reply ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt:86]: Ah! Then, yes, you're having the same problems unless you remember a `Flow.flowWithLifecycle` as you can read here: https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda

Relevant code snippet: https://gist.github.com/manuelvicnt/6629b53e33b97c127508227ed6e0bee4#file-flowwithlifecyclecompose-kt

File: common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt (hunk 2 above)
Comment (manuelvicnt) [reply common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt:36]: The following code buffers UiErrors in the channel when there are no collectors

```
private val _pendingErrors = Channel<UiError>(3, DROP_OLDEST)
val pendingErrors = _navigationEvents.receiveAsFlow().shareIn(externalScope, WhileSubscribed())
```

File: ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt (hunk 4 above)
Comment (chrisbanes) [reply ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt]: Migrated to `flowWithLifecycle()` 👍

File: ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt (hunk 3 above)
Comment (chrisbanes) [reply ui-episodedetails/src/main/java/app/tivi/episodedetails/EpisodeDetailsViewModel.kt]: Good call, removed.

File: common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt (hunk 2 above)
Comment (chrisbanes) [reply common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt]: Thanks, went back to using a Channel. Wasn't sure that we need the `shareIn()`. PTAL

File: common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt (hunk 1 above)
Comment (chrisbanes) [reply common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt]: Added.

**Review by manuelvicnt — APPROVED** [review]

File: common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt (hunk 2 above)
Comment (manuelvicnt) [reply common-ui-view/src/main/java/app/tivi/ui/SnackbarManager.kt]: Looks good 👍  thank you!

File: data-android/src/main/java/app/tivi/data/TiviRoomDatabase.kt [hunk 5]
```diff
@@ -62,7 +62,6 @@ import app.tivi.data.views.FollowedShowsWatchStats
     ],
     version = 26,
     autoMigrations = [
-        AutoMigration(from = 23, to = 24),
```
Comment (chrisbanes) [inline data-android/src/main/java/app/tivi/data/TiviRoomDatabase.kt:65]: No idea why the build doesn't like it. It's not a big issue to remove this anyway.

File: common-ui-compose/src/main/java/app/tivi/common/compose/Flow.kt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            lifecycle = lifecycleOwner.lifecycle,
+            minActiveState = minActiveState
+        ).collect {
+            value = it
+        }
+    }
+}
+
+@Composable
+inline fun <T> Flow<T>.flowWithLocalLifecycle(
```
Comment (manuelvicnt) [inline common-ui-compose/src/main/java/app/tivi/common/compose/Flow.kt:47]: Does this need to be remembered for paging? If so, you might want to call it `inline fun rememberFlowWithLocalLifecycle(flow)

Comment (chrisbanes) [reply common-ui-compose/src/main/java/app/tivi/common/compose/Flow.kt:47]: It actually works fine right now, but I like the look of `rememberFlowWithLifecycle` so migrated all calls to it

**Pull request merged**